frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx

```diff
@@ -0,0 +1,339 @@
+
+import { FileUploadClient } from "@streamlit/lib/src/FileUploadClient"
+import { WidgetStateManager } from "@streamlit/lib/src/WidgetStateManager"
+import { AudioInput as AudioInputProto } from "@streamlit/lib/src/proto"
+import Toolbar, {
+  ToolbarAction,
+} from "@streamlit/lib/src/components/shared/Toolbar"
+import { EmotionTheme } from "@streamlit/lib/src/theme"
+import {
+  isNullOrUndefined,
+  labelVisibilityProtoValueToEnum,
+  notNullOrUndefined,
+} from "@streamlit/lib/src/util/utils"
+import { WidgetLabel } from "@streamlit/lib/src/components/widgets/BaseWidget"
+
+import { uploadFiles } from "./uploadFiles"
+import {
+  StyledAudioInputContainerDiv,
+  StyledWaveformContainerDiv,
+  StyledWaveformInnerDiv,
+  StyledWaveformTimeCode,
+  StyledWaveSurferDiv,
+} from "./styled-components"
+import NoMicPermissions from "./NoMicPermissions"
+import Placeholder from "./Placeholder"
+import {
+  BAR_GAP,
+  BAR_RADIUS,
+  BAR_WIDTH,
+  CURSOR_WIDTH,
+  STARTING_TIME_STRING,
+  WAVEFORM_HEIGHT,
+  WAVEFORM_PADDING,
+} from "./constants"
+import formatTime from "./formatTime"
+import ActionButton from "./ActionButton"
+
+interface Props {
+  element: AudioInputProto
+  uploadClient: FileUploadClient
+  widgetMgr: WidgetStateManager
+  theme: EmotionTheme
+}
+
+const AudioInput: React.FC<Props> = ({
+  element,
+  uploadClient,
+  widgetMgr,
+  theme,
+}): ReactElement => {
+  const [wavesurfer, setWavesurfer] = useState<WaveSurfer | null>(null)
+  const waveSurferRef = React.useRef<HTMLDivElement | null>(null)
+  const [deleteFileUrl, setDeleteFileUrl] = useState<string | null>(null)
+  const [recordPlugin, setRecordPlugin] = useState<RecordPlugin | null>(null)
+  // to eventually show the user the available audio devices
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [availableAudioDevices, setAvailableAudioDevices] = useState<
+    MediaDeviceInfo[]
+  >([])
+  const [activeAudioDeviceId, setActiveAudioDeviceId] = useState<
+    string | null
+  >(null)
+  const [recordingUrl, setRecordingUrl] = useState<string | null>(null)
+  const [, setRerender] = useState(0)
+  const forceRerender = (): void => {
+    setRerender(prev => prev + 1)
+  }
+  const [progressTime, setProgressTime] = useState(STARTING_TIME_STRING)
+  const [recordingTime, setRecordingTime] = useState(STARTING_TIME_STRING)
+  const [shouldUpdatePlaybackTime, setShouldUpdatePlaybackTime] =
+    useState(false)
+  const [hasNoMicPermissions, setHasNoMicPermissions] = useState(false)
+
+  const uploadTheFile = useCallback(
+    (file: File) => {
+      uploadFiles({
+        files: [file],
+        uploadClient,
+        widgetMgr,
+        widgetInfo: element,
+      }).then(({ successfulUploads }) => {
+        const upload = successfulUploads[0]
+        if (upload && upload.fileUrl.deleteUrl) {
+          setDeleteFileUrl(upload.fileUrl.deleteUrl)
+        }
+      })
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
**question:** Why are we disabling this? Will it lead to stale values in the callback closure?

Reply:
I didn't get enough time to look into _why_ this is happening, but it seems that element is changing much more than I expected. Luckily, we're just using the element for its [WidgetInfo](https://github.com/streamlit/streamlit/blob/nico/audio-input/frontend/lib/src/WidgetStateManager.ts#L43) so we don't really care about those updates. 

We could probably clean this up by extracting that widget info and including that new object in the deps array? wdyt?

Oh yeah and to why this was a problem: the `intitializeWavesurfer` depends on this function, and it would cause that to re-run and that re-running causes wavesurfer to be reinitialized and loses all the local state. 😞

Reply:
> Will it lead to stale values in the callback closure?

No, but I think we can make this more self documenting.